File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.java
class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        int n = queries.length;
        int[] ans = new int[n];
        Map<Integer,Integer> cntColor = new HashMap<>();
        Map<Integer,Integer> trackBallColor = new HashMap<>();
        for(int i = 0; i<n; i++){
            int ball = queries[i][0];
            int color = queries[i][1];
            if(trackBallColor.containsKey(ball)){// color already present
                int prevColor = trackBallColor.get(ball);
                trackBallColor.put(ball, color);
                int cnt = cntColor.get(prevColor)-1;
                if(cnt == 0){
                    cntColor.remove(prevColor);
                }else{
                    cntColor.put(prevColor, cnt);
                }
            }
            trackBallColor.put(ball,color);
            int cnt = cntColor.getOrDefault(color, 0) + 1;
            cntColor.put(color, cnt);
            ans[i] = cntColor.size();
        }
        return ans;
    }
}

File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        vector<int> ans;
        int diffColor = 0;
        int n = queries.size();
        unordered_map<int,int> mpBallColor;
        unordered_map<int, int> cntColor;
        for(int i = 0; i<n; i++){
            int ball = queries[i][0];
            int color = queries[i][1];
            if(mpBallColor.find(ball)  != mpBallColor.end()){ // ball is already present
                int prevColor = mpBallColor[ball];
                mpBallColor[ball] = color;
                cntColor[prevColor]--;

                // if color == 0 , ek color kam ho gya then remove the color from map
                if(cntColor[prevColor] == 0)
                    cntColor.erase(prevColor);
            }
            mpBallColor[ball] = color;
            cntColor[color]++;
            ans.push_back(cntColor.size());
        }
        return ans;
    }
};